### PR TITLE
Poverty example

### PR DESCRIPTION
**Pull request by Zathimo**
## :memo: Changelog

- [x] new regression model
- [x] poverty data module
- [x] example script to run the regression model on poverty
- [ ]  easy access to images is yet to be implemented

## :white_check_mark: Checklist

- [ ] Lint and tests pass locally with my changes
- [x] I've added necessary documentation

### CODE REVIEW

File: malpolon/data/datasets/poverty_dataset.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+        self.dataframe = dataframe
+        self.dataframe_train = dataframe[dataframe['subset'] == 'train']
+        self.dataframe_val = dataframe[dataframe['subset'] == 'val']
+        self.dataframe_test = dataframe[dataframe['subset'] == 'test']
+        self.tif_dir = dataset_path + tif_dir
+        self.train_batch_size = train_batch_size
+        self.inference_batch_size = inference_batch_size
+        self.dict_normalize = json.load(open('examples/poverty/mean_std_normalize.json', 'r'))
+        self.num_workers = num_workers
+        self.task = 'regression'
```
Comment (tlarcher) [inline malpolon/data/datasets/poverty_dataset.py:54]: This parameter is expected to be registered in the config file, however here it is hard coded. Consider adding a 'task' parameter in your class __init__() method; or check whether the task parameter exists in **kwargs (taking its value if True) before assigning 'regression'

File: malpolon/data/datasets/poverty_dataset.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+            dataset_path (str): path to the dataset
+            labels_name (str): name of the csv file containing the labels
+            train_batch_size (int): batch size for training
+            inference_batch_size (int): batch size for inference
+            num_workers (int): number of workers for the DataLoader
+            fold (int): fold to use for training
+            transform (torchvision.transforms): transform to apply to the data"""
+
+        super().__init__()
+        dataframe = pd.read_csv(dataset_path + labels_name)
```
Comment (tlarcher) [inline malpolon/data/datasets/poverty_dataset.py:44]: Consider encapsulating your strings in `pathlib.Path` objects to avoid '/' (slash) sensitiveness.
e.g.: 
```
from pathlib import Path
dataframe = pd.read_csv(Path(dataset_path) / Path(labels_name))
```

File: malpolon/data/datasets/poverty_dataset.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+        self.dataframe = dataframe
+        self.dataframe_train = dataframe[dataframe['subset'] == 'train']
+        self.dataframe_val = dataframe[dataframe['subset'] == 'val']
+        self.dataframe_test = dataframe[dataframe['subset'] == 'test']
+        self.tif_dir = dataset_path + tif_dir
```
Comment (tlarcher) [inline malpolon/data/datasets/poverty_dataset.py:49]: Consider encapsulating your strings in `pathlib.Path` objects to avoid '/' (slash) sensitiveness.
e.g.: 
```
from pathlib import Path
self.tif_dir = Path(dataset_path) + Path(tif_dir)
```

File: malpolon/data/datasets/poverty_dataset.py (hunk 2 above)
Comment (tlarcher) [inline malpolon/data/datasets/poverty_dataset.py:44]: Consider encapsulating your strings in `pathlib.Path` objects to avoid '/' (slash) sensitiveness.
e.g.: 
```
from pathlib import Path
dataframe = pd.read_csv(Path(dataset_path) / Path(labels_name))
```

File: malpolon/data/datasets/poverty_dataset.py (hunk 3 above)
Comment (tlarcher) [inline malpolon/data/datasets/poverty_dataset.py:49]: Consider encapsulating your strings in `pathlib.Path` objects to avoid '/' (slash) sensitiveness.
e.g.: 
```
from pathlib import Path
self.tif_dir = Path(dataset_path) / Path(tif_dir)
```

File: malpolon/data/datasets/poverty_dataset.py (hunk 1 above)
Comment (tlarcher) [inline malpolon/data/datasets/poverty_dataset.py:54]: The task argument is expected to be registered in an example config file. However here it is hardcoded.

Consider either adding the 'task' argument as input of your __init__() method; or checking in **kwargs if such and argument exists (taking priority over the default value 'regression')

File: malpolon/data/datasets/poverty_dataset.py [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    def get_dataset(self, split: str, **kwargs) -> Dataset:
+        if split == 'train':
+            dataset = MSDataset(self.dataframe_train, self.tif_dir, transform=self.train_transform())
+        elif split == 'val':
+            dataset = MSDataset(self.dataframe_val, self.tif_dir, transform=self.train_transform())
+        elif split == 'test':
+            dataset = MSDataset(self.dataframe_test, self.tif_dir, transform=self.test_transform())
+        return dataset
+
+    def train_dataloader(self):
```
Comment (tlarcher) [inline malpolon/data/datasets/poverty_dataset.py:79]: As your code is, it is only useful to re-define train/val/test_dataloaders to use `persistent_workers=True`, which is relevant if your epochs are fast and you want to make your training faster by avoiding to re-create workers at each epoch. This is done at the cost of memory.

If this is the intended use, ignore this comment. In the contrary case, you can remove altogether the re-defined train/val/test_dataloaders

File: malpolon/models/standard_prediction_systems.py [hunk 5]
```diff
@@ -160,14 +163,18 @@ def _step(
         y_hat = self(x)
 
         loss = self.loss(y_hat, self._cast_type_to_loss(y))  # Shape mismatch for binary: need to 'y = y.unsqueeze(1)' (or use .reshape(2)) to cast from [2] to [2,1] and cast y to float with .float()
-        self.log(f"loss/{split}", loss, **log_kwargs)
+        self.log(f"loss_{split}", loss, **log_kwargs)
```
Comment (tlarcher) [inline malpolon/models/standard_prediction_systems.py:166]: The `/` (slash) character was chosen as a separator between the quantity to monitor and the split, because Tensorboard recognizes it and gathers all quantity-related measurements in a single section. This is no the case with other characters.
Hence, `/` (slash) should be kept as a separator.

Screenshot 2024-11-08 at 20-03-48 TensorBoard
Screenshot 2024-11-08 at 20-03-31 TensorBoard
Screenshot 2024-11-08 at 20-03-19 TensorBoard

File: malpolon/models/standard_prediction_systems.py [hunk 6]
```diff
@@ -160,14 +163,18 @@ def _step(
...
-                score = metric_func['callable'](y_hat, y, **metric_func['kwargs'])
+                if metric_func['kwargs']:
+
+                    score = metric_func['callable'](y_hat, y, **metric_func['kwargs'])
+                else:
+                    score = metric_func['callable'](y_hat, y)
             else:
                 score = metric_func(y_hat, y)
-            self.log(f"{metric_name}/{split}", score, **log_kwargs)
+            self.log(f"{metric_name}_{split}", score, **log_kwargs)
```
Comment (tlarcher) [inline malpolon/models/standard_prediction_systems.py:177]: Same as https://github.com/plantnet/malpolon/pull/66/files#r1834916272

File: malpolon/models/standard_prediction_systems.py (hunk 5 above)
Comment (tlarcher) [reply malpolon/models/standard_prediction_systems.py:166]: Changing that character would also mean updating every example

File: malpolon/models/standard_prediction_systems.py [hunk 7]
```diff
@@ -160,14 +163,18 @@ def _step(
         y_hat = self(x)
 
         loss = self.loss(y_hat, self._cast_type_to_loss(y))  # Shape mismatch for binary: need to 'y = y.unsqueeze(1)' (or use .reshape(2)) to cast from [2] to [2,1] and cast y to float with .float()
-        self.log(f"loss/{split}", loss, **log_kwargs)
+        self.log(f"loss_{split}", loss, **log_kwargs)
 
         for metric_name, metric_func in self.metrics.items():
             if isinstance(metric_func, dict):
-                score = metric_func['callable'](y_hat, y, **metric_func['kwargs'])
+                if metric_func['kwargs']:
```
Comment (tlarcher) [inline malpolon/models/standard_prediction_systems.py:170]: Indeed, metric kwargs are expected to be added in the config file; however, when the `kwargs` is empty, calling metric_func with kwargs will trigger a None-type error.

I suggest you replace that if condition by the following line instead:
`score = metric_func['callable'](y_hat, y, metric_func.get('kwargs', {}))`
replacing the old line 167.

This is a good addition, thx!

File: malpolon/models/standard_prediction_systems.py [hunk 8]
```diff
@@ -348,6 +355,7 @@ def predict_point(
         """
         device = 'cuda' if torch.cuda.is_available() else 'cpu'
         self.model.to(device)
+        data = data.to(device)
```
Comment (tlarcher) [inline malpolon/models/standard_prediction_systems.py:358]: `data` isn't guaranteed to be a Tensor since this method doesn't require it to be (intentional behavior since the method should be callable by hand on any data type).
If `data` isn't a Tensor, the calling of `.to()` will fail and crash.
The data (or its elements) are casted to `device` anyway in the following lines.

=> Recommendation: delete this line

File: malpolon/models/standard_prediction_systems.py [hunk 9]
```diff
@@ -448,4 +456,92 @@ def __init__(
...
+        self.momentum = momentum
+        self.nesterov = nesterov
+
+        model = check_model(model)
+
+        optimizer = torch.optim.AdamW(model.parameters(),
+                                      lr=lr,
+                                      weight_decay=weight_decay)
+
+        print(optimizer)
```
Comment (tlarcher) [inline malpolon/models/standard_prediction_systems.py:519]: Remove debug print

File: malpolon/models/standard_prediction_systems.py [hunk 10]
```diff
@@ -448,4 +456,92 @@ def __init__(
...
+
+        if 'regression' in task:
```
Comment (tlarcher) [inline malpolon/models/standard_prediction_systems.py:521]: No default value for `loss`. Risk of error if `task` doesn't contain 'regression' string as `loss` will be None. The error is caught by malpolon.models.utils.check_loss() but I advise handling a 'default loss value' case

File: malpolon/models/standard_prediction_systems.py [hunk 11]
```diff
@@ -448,4 +456,92 @@ def __init__(
...
+
+
+        self.lr = lr
+        self.weight_decay = weight_decay
+        self.momentum = momentum
+        self.nesterov = nesterov
+
+        model = check_model(model)
+
+        optimizer = torch.optim.AdamW(model.parameters(),
```
Comment (tlarcher) [inline malpolon/models/standard_prediction_systems.py:515]: INFO: optimizers instantiation have changed behavior since update v2.1.0 and is now a user choice (via config file.

This is still 100% compatible with the update, but consider adding a `optimizer` input argument and nesting your optimizer variable instantiation inside the condition `if optimizer is None:`

(see malpolon.models.standard_prediction_systems.ClassificationSystem)

File: malpolon/models/standard_prediction_systems.py [hunk 12]
```diff
@@ -448,4 +456,92 @@ def __init__(
...
+
+        if metrics is None:
+            metrics = {
+                "regression_R2score": {'callable': Fmetrics.regression.r2_score,
+                                       'kwargs': {}}
+            }
+
         super().__init__(model, loss, optimizer, metrics=metrics)
+
+    def configure_optimizers(self):
```
Comment (tlarcher) [inline malpolon/models/standard_prediction_systems.py:532]: INFO: since the v2.1.0 update, this method doesn't need to be re-defined unless you want to impose a fixed optimizer/scheduler as both of these objects can be defined by the user in his experiment config file

File: malpolon/models/utils.py [hunk 13]
```diff
@@ -11,6 +11,7 @@
 from pathlib import Path
 from typing import Mapping, Union
 
+import torchmetrics
```
Comment (tlarcher) [inline malpolon/models/utils.py:14]: Import not used ?

**Review by tlarcher — CHANGES_REQUESTED** [review]
Reviewed the engine side of your PR as requested. Couple of things to address. I'll let you remove the example/dataset part for now.
When everything is addressed I'll merge this in a new branch which will wait for you examples when they are ready.

If you want, you can adapt the (optimizer + config file) parts in accordance to the v2.1.0 update but that isn't necessary.

File: malpolon/models/standard_prediction_systems.py [hunk 14]
```diff
@@ -448,4 +456,92 @@ def __init__(
                              'kwargs': {}}
             }
 
+        super().__init__(model, loss, optimizer, metrics)
+
+
+class RegressionSystem(GenericPredictionSystem):
+    """Regression task class."""
+    def __init__(
```
Comment (tlarcher) [inline malpolon/models/standard_prediction_systems.py:464]: Some arguments  (optimizer kwargs) are not used because incompatible with the proposed default optimizer. I suggest replacing them with the default optimizer's; or simply removing them.

Please update the docstring accordingly too

File: malpolon/models/standard_prediction_systems.py (hunk 5 above)
Comment (Zathimo) [reply malpolon/models/standard_prediction_systems.py]: I didn't think of that. I undid it

File: malpolon/data/datasets/poverty_dataset.py (hunk 2 above)
Comment (Zathimo) [reply malpolon/data/datasets/poverty_dataset.py]: As we discussed, I removed the dataset from the PR.
I did make the chage though, thanks for the advice!

File: malpolon/data/datasets/poverty_dataset.py (hunk 3 above)
Comment (Zathimo) [reply malpolon/data/datasets/poverty_dataset.py]: As we discussed, I removed the dataset from the PR.
I did make the chage though, thanks for the advice!

File: malpolon/data/datasets/poverty_dataset.py (hunk 2 above)
Comment (Zathimo) [reply malpolon/data/datasets/poverty_dataset.py]: As we discussed, I removed the dataset from the PR.
I did make the chage though, thanks for the advice!

File: malpolon/data/datasets/poverty_dataset.py (hunk 3 above)
Comment (Zathimo) [reply malpolon/data/datasets/poverty_dataset.py]: As we discussed, I removed the dataset from the PR.
I did make the chage though, thanks for the advice!

File: malpolon/data/datasets/poverty_dataset.py (hunk 4 above)
Comment (Zathimo) [reply malpolon/data/datasets/poverty_dataset.py]: As we discussed, I removed the dataset from the PR.
I did make the chage though, thanks for the advice!

File: malpolon/models/standard_prediction_systems.py (hunk 6 above)
Comment (Zathimo) [reply malpolon/models/standard_prediction_systems.py]: done

File: malpolon/models/utils.py (hunk 13 above)
Comment (Zathimo) [reply malpolon/models/utils.py]: withdrawn

File: malpolon/models/standard_prediction_systems.py (hunk 12 above)
Comment (Zathimo) [reply malpolon/models/standard_prediction_systems.py]: withdrawn

File: malpolon/models/standard_prediction_systems.py (hunk 14 above)
Comment (Zathimo) [reply malpolon/models/standard_prediction_systems.py:457]: removed the kwargs
